sklearn/cluster/_hdbscan/_reachability.pyx

```diff
@@ -0,0 +1,103 @@
+# mutual reachability distance compiutations
+# Authors: Leland McInnes
+# License: 3-clause BSD
+
+import numpy as np
+
+cimport numpy as np
+
+import gc
+
+from scipy.sparse import lil_matrix as sparse_matrix
+from scipy.spatial.distance import pdist, squareform
+
+from sklearn.neighbors import BallTree, KDTree
+
+
+def mutual_reachability(distance_matrix, min_points=5, alpha=None):
+    """Compute the weighted adjacency matrix of the mutual reachability
+    graph of a distance matrix.
+
+    Parameters
+    ----------
+    distance_matrix : ndarray, shape (n_samples, n_samples)
+        Array of distances between samples.
+
+    min_points : int, default=5
+        The number of points in a neighbourhood for a point to be considered
+        a core point.
+
+    alpha : float, default=None
+        A distance scaling parameter as used in robust single linkage. This
+        divides the distances when calculating mutual reachability.
+
+    Returns
+    -------
+    mututal_reachability: ndarray, shape (n_samples, n_samples)
+        Weighted adjacency matrix of the mutual reachability graph.
+
+    References
+    ----------
+    .. [1] Campello, R. J., Moulavi, D., & Sander, J. (2013, April).
+       Density-based clustering based on hierarchical density estimates.
+       In Pacific-Asia Conference on Knowledge Discovery and Data Mining
+       (pp. 160-172). Springer Berlin Heidelberg.
+    """
+    size = distance_matrix.shape[0]
+    min_points = min(size - 1, min_points)
+    try:
+        core_distances = np.partition(distance_matrix,
+                                      min_points,
+                                      axis=0)[min_points]
+    except AttributeError:
+        core_distances = np.sort(distance_matrix,
+                                 axis=0)[min_points]
+
+    if alpha is not None:
+        distance_matrix = distance_matrix / alpha
+
+    stage1 = np.where(core_distances > distance_matrix,
+                      core_distances, distance_matrix)
+    result = np.where(core_distances > stage1.T,
+                      core_distances.T, stage1.T).T
+    return result
+
+
+cpdef sparse_mutual_reachability(object lil_matrix, np.intp_t min_points=5,
+                                 float alpha=1.0, float max_dist=0.):
```

Comment:
since the LIL matrix will be of type `np.float64`, I think that we should define `max_dist` as a `double`. Otherwise, we are going to make some implicit casting somewhere for the comparaison.